sklearn/tree/_splitter.pyx

```diff
@@ -220,222 +221,178 @@ cdef class Splitter:
-
-        # Sample up to max_features without replacement using a
-        # Fisher-Yates-based algorithm (using the local variables `f_i` and
-        # `f_j` to compute a permutation of the `features` array).
-        #
-        # Skip the CPU intensive evaluation of the impurity criterion for
-        # features that were already detected as constant (hence not suitable
-        # for good splitting) by ancestor nodes and save the information on
-        # newly discovered constant features to spare computation on descendant
-        # nodes.
-        while (f_i > n_total_constants and  # Stop early if remaining features
-                                            # are constant
-                (n_visited_features < max_features or
-                 # At least one drawn features must be non constant
-                 n_visited_features <= n_found_constants + n_drawn_constants)):
-
-            n_visited_features += 1
-
-            # Loop invariant: elements of features in
-            # - [:n_drawn_constant[ holds drawn and known constant features;
-            # - [n_drawn_constant:n_known_constant[ holds known constant
-            #   features that haven't been drawn yet;
-            # - [n_known_constant:n_total_constant[ holds newly found constant
-            #   features;
-            # - [n_total_constant:f_i[ holds features that haven't been drawn
-            #   yet and aren't constant apriori.
-            # - [f_i:n_features[ holds features that have been drawn
-            #   and aren't constant.
-
-            # Draw a feature at random
-            f_j = rand_int(n_drawn_constants, f_i - n_found_constants,
-                           random_state)
-
-            if f_j < n_known_constants:
-                # f_j in the interval [n_drawn_constants, n_known_constants[
-                features[n_drawn_constants], features[f_j] = features[f_j], features[n_drawn_constants]
-
-                n_drawn_constants += 1
+ctypedef fused DataSplitterFused:
+    DenseSplitter
+    SparseSplitter
+
+cdef inline int node_split_best(
+    Splitter splitter,
+    DataSplitterFused data_splitter,
+    Criterion criterion,
+    double impurity,
+    SplitRecord* split,
+    SIZE_t* n_constant_features,
+) nogil except -1:
+    """Find the best split on node samples[start:end]
+
+    Returns -1 in case of failure to allocate memory (and raise MemoryError)
+    or 0 otherwise.
+    """
+    # Find the best split
+    cdef SIZE_t start = splitter.start
+    cdef SIZE_t end = splitter.end
+
+    cdef SIZE_t[::1] features = splitter.features
+    cdef SIZE_t[::1] constant_features = splitter.constant_features
+    cdef SIZE_t n_features = splitter.n_features
+
+    cdef DTYPE_t[::1] Xf = splitter.feature_values
+    cdef SIZE_t max_features = splitter.max_features
+    cdef SIZE_t min_samples_leaf = splitter.min_samples_leaf
+    cdef double min_weight_leaf = splitter.min_weight_leaf
+    cdef UINT32_t* random_state = &splitter.rand_r_state
```

Comment:
@thomasjpfan I am sure that you already try to refactor in a different manner. The current refactoring is probably the easiest and with the smallest diff.

But this part here makes me wonder. We do not really use the `splitter`. In some way, those could be attached to the `data_splitter` directly. Thus, instead of creating the `DenseSplitter` or `SparseSplitter` in the specialized splitter, one could make a composition of the classes. In someway, `DenseSplitter`, `SparseSplitter`, `RandomSplitter` and `BestSplitter` would just be some mixin and with some multiple inheritances you would have something like:

```python
class BestDenseSplitter(DenseSplitter, BestSplitter):
    pass
```

Of course, I am sure that you already look at this and I am not even sure how the Cython classes would behave with such multiple inheritances.

Reply:
This might be worth adding a small comment before declaring the 4 specialized classes to make it explicit why this design was chosen such that nobody tries to reinvent the wheel. So I am fine with the current design then.